File: Leet 23/solution.py
class ListNode:
    def __init__(self, val: int):
        self.val: int = val
        self.next: ListNode|None = None


class Solution:
    def list_to_listnode(self, values: list[int]):
        head: ListNode|None = None
        tail: ListNode|None = None

        for val in values:
            if head == None:
                head = ListNode(val)
                tail = head
            elif tail:
                tail.next = ListNode(val)
                tail = tail.next

        return head

    def mergeKLists(self, lists: list[ListNode|None]) -> ListNode|None:
        if not lists:
            return None

        values: list[int] = []
        for current in lists:
            while current:
                values.append(current.val)
                current = current.next
        values.sort()

        return self.list_to_listnode(values)


def main() -> None:
    print('23. Merge k Sorted Lists')

    solution = Solution()

    case1 = [
        solution.list_to_listnode([1,4,5]),
        solution.list_to_listnode([1,3,4]),
        solution.list_to_listnode([2,6]),
    ]
    result = solution.mergeKLists(case1)
    while result:
        print(result.val)
        result = result.next


if __name__ == '__main__':
    main()

File: Leet 23/solution.cs
public class ListNode 
{
    public int val;
    public ListNode? next;
    public ListNode(int val=0, ListNode? next=null) 
    {
        this.val = val;
        this.next = next;
    }
}

public class Solution 
{
    public ListNode? MergeKLists(ListNode?[] lists) 
    {
        List<int> values = [];
        foreach(var list in lists)
        {
            var current = list;
            while(current != null)
            {
                values.Add(current.val);
                current = current.next;
            }
        }
        values.Sort();

        ListNode? head = null;
        ListNode? tail = null;
        foreach(var value in values)
        {
            if(head == null)
            {
                head = new ListNode(value);
                tail = head;
            }
            else if(tail != null)
            {
                tail.next = new ListNode(value);
                tail = tail.next;
            }
        }
        return head;
    }
}

static class Program
{
    private static ListNode? ArrayToListNode(int[] values)
    {
        ListNode? head = null;
        ListNode? tail = null;
        foreach(var value in values)
        {
            if(head == null)
            {
                head = new ListNode(value);
                tail = head;
            }
            else if(tail != null)
            {
                tail.next = new ListNode(value);
                tail = tail.next;
            }
        }
        return head;
    }

    private static void Main()
    {
        Solution solution = new();
        ListNode?[] lists = [
            ArrayToListNode([1,4,5]),
            ArrayToListNode([1,3,4]),
            ArrayToListNode([2,6]),
        ];
        var head = solution.MergeKLists(lists);
        Console.Write("[ ");
        while(head != null)
        {
            Console.Write($" {head.val} ");
            head = head.next;
        }
        Console.WriteLine(" ]");
    }
}